### PR TITLE
Log4 j j unit

### PR DESCRIPTION
**Pull request by yurachel**

### CODE REVIEW

File: src/main/java/by/yurachel/web_app/repository/PhoneRepository.java [hunk 1]
```diff
@@ -46,8 +48,47 @@ public void changePhoneParam(Phone newPhone, String oldPhoneName) {
...
+       boolean isRemove =  phones.removeIf(phone -> phone.getId() == id);
+       if(!isRemove){
+           throw new IllegalArgumentException("The phone wasn't found");
+       }
 
     }
+
+    public Phone findPhoneById(int id) {
+        Phone findPhone = new Phone();
+        for (Phone phone : phones) {
```
Comment (anikosss) [inline src/main/java/by/yurachel/web_app/repository/PhoneRepository.java:61]: Use stream here

File: src/main/java/by/yurachel/web_app/repository/PhoneRepository.java [hunk 2]
```diff
@@ -46,8 +48,47 @@ public void changePhoneParam(Phone newPhone, String oldPhoneName) {
...
+                break;
+            }
+        }
+        return findPhone;
+    }
+
+
+    public int findPhoneIDByName(String name) {
+        Integer findIdPhone = null;
+        for (Phone phone : phones) {
```
Comment (anikosss) [inline src/main/java/by/yurachel/web_app/repository/PhoneRepository.java:73]: Use stream here

File: src/main/java/by/yurachel/web_app/repository/PhoneRepository.java [hunk 3]
```diff
@@ -46,8 +48,47 @@ public void changePhoneParam(Phone newPhone, String oldPhoneName) {
...
+            if (phone.getName().equals(name)) {
+                findIdPhone = phone.getId();
+                break;
+            }
+        }
+        if (findIdPhone == null) {
+            throw new NullPointerException("Such phone wasn't found");
```
Comment (anikosss) [inline src/main/java/by/yurachel/web_app/repository/PhoneRepository.java:80]: **Never throw NPE manually**, there should be any other exception (better approach is custom exception), or at least IllegalArgumentException

File: src/main/java/by/yurachel/web_app/repository/PhoneRepository.java [hunk 4]
```diff
@@ -3,24 +3,26 @@
...
 import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
 
 public class PhoneRepository {
     private static final PhoneRepository INSTANCE = new PhoneRepository();
-
     private final List<Phone> phones = new ArrayList<>();
+    public static final List<Integer> ID = new ArrayList<>(Arrays.asList(1, 2, 3, 4, 5));
```
Comment (anikosss) [inline src/main/java/by/yurachel/web_app/repository/PhoneRepository.java:13]: You are editing this field, so it should be final.
Also it should be private with getters/setters (or other access methods)

File: src/main/resources/log4j2.xml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Configuration status="WARN">
+
+    <Appenders>
+        <Console name="Console" target="SYSTEM_OUT">
+            <PatternLayout pattern="%d{HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n"/>
+        </Console>
+        <RollingFile name="RollingFile" fileName="logs/app.log"
+                     filePattern="logs/$${date:yyyy-MM}/app-%d{MM-dd-yyyy}-%i.log.gz">
```
Comment (anikosss) [inline src/main/resources/log4j2.xml:9]: There should be path to Tomcat /logs directory, it should begin from "../logs" as default directory for application is Tomcat/bin

File: src/test/java/by/yurachel/web_app/servlet/AddPhoneServletTest.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    void setUp() {
+        addPhoneServlet = new AddPhoneServlet();
+        pr = mock(PhoneRepository.class);
+        capPhoneRep = Mockito.mockStatic(PhoneRepository.class);
+    }
+
+
+    @Test
+    void doGetTest() throws ServletException, IOException {
+        HttpServletRequest request = mock(HttpServletRequest.class);
```
Comment (anikosss) [inline src/test/java/by/yurachel/web_app/servlet/AddPhoneServletTest.java:37]: Reqeust/response/dispatcher mocks are duplicated in every method, move it to:
1) init method
2) use @Mock annotation for fields
3) create abstract test class

File: src/test/java/by/yurachel/web_app/servlet/AddPhoneServletTest.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        addPhoneServlet.doPost(request, response);
+
+    }
+
+    @Test
+    void when_Param_Name_Is_null() throws IOException {
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        HttpServletResponse response = mock(HttpServletResponse.class);
+        when(request.getParameter("name")).thenReturn(null);
+        try {
```
Comment (anikosss) [inline src/test/java/by/yurachel/web_app/servlet/AddPhoneServletTest.java:76]: Use JUnit "assertThrows" expression to test exception case

File: src/main/java/by/yurachel/web_app/dao/ConnectorDB.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package by.yurachel.web_app.dao;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.util.Properties;
+
+public class ConnectorDB {
```
Comment (danikeichanka) [inline src/main/java/by/yurachel/web_app/dao/ConnectorDB.java:8]: Should be singleton, or just move this method to AbstractDAO

File: src/main/java/by/yurachel/web_app/dao/AbstractDAO.java [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+    public abstract T findEntityById(int id);
+
+    public abstract boolean remove(T object);
+
+    public abstract boolean remove(int id);
+
+    public abstract boolean update(T object);
+
+    public abstract boolean addEntity(T object);
```
Comment (danikeichanka) [inline src/main/java/by/yurachel/web_app/dao/AbstractDAO.java:16]: Should return created entity or created entity ID

File: src/main/java/by/yurachel/web_app/dao/AbstractDAO.java [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+public abstract class AbstractDAO<T> {
+    public abstract List<T> findAll();
+
+    public abstract T findEntityById(int id);
+
+    public abstract boolean remove(T object);
+
+    public abstract boolean remove(int id);
+
+    public abstract boolean update(T object);
```
Comment (danikeichanka) [inline src/main/java/by/yurachel/web_app/dao/AbstractDAO.java:14]: Should return updated entity.

File: src/main/java/by/yurachel/web_app/dao/PhoneListDAO.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+            while (rs.next()) {
+                long id = rs.getLong(1);
+                String name = rs.getString(2);
+                double price = rs.getDouble(3);
+                String processor = rs.getString(4);
+                Phone phone = new Phone(id, name, price, processor);
+                phones.add(phone);
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
```
Comment (danikeichanka) [inline src/main/java/by/yurachel/web_app/dao/PhoneListDAO.java:34]: Don't call prinsStackTrace, should be deligated to logger

File: src/main/java/by/yurachel/web_app/filter/EncodingFilter.java [hunk 12]
```diff
@@ -6,15 +6,19 @@
...
 import jakarta.servlet.http.HttpFilter;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 
 @WebFilter(filterName = "encoding", urlPatterns = "/*")
 public class EncodingFilter extends HttpFilter {
+    private static final Logger ROOT_LOGGER = LogManager.getRootLogger();
```
Comment (danikeichanka) [inline src/main/java/by/yurachel/web_app/filter/EncodingFilter.java:17]: Call this field just LOGGER, no need to specidy ROOT prefix (it was on lesson just as an example that we can inject view logger to 1 class)

Logger should be retrieved by calling another method:

_LogManager.getLogger(EncodingFilter.class)_

File: src/main/java/by/yurachel/web_app/repository/PhoneRepository.java [hunk 13]
```diff
@@ -1,33 +1,42 @@
...
+import by.yurachel.web_app.dao.PhoneListDAO;
 import by.yurachel.web_app.entity.Phone;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
-import java.util.ArrayList;
-import java.util.List;
+import java.util.*;
 
 public class PhoneRepository {
```
Comment (danikeichanka) [inline src/main/java/by/yurachel/web_app/repository/PhoneRepository.java:10]: This is redundant class as we added DAO and entities are now retrieved from the DB

File: src/test/java/by/yurachel/web_app/servlet/AddPhoneServletTest.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+    @Test
+    void testDoPost() throws IOException {
+        when(phoneRepository.maxPhoneID()).thenReturn(5L);
+        when(request.getParameter("name")).thenReturn("anyString()");
+        when(request.getParameter("price")).thenReturn("2");
+        when(request.getParameter("processor")).thenReturn("snap");
+        doNothing().when(phoneRepository).addPhone(any(Phone.class));
+        addPhoneServlet.doPost(request, response);
+        verify(phoneRepository).addPhone(any(Phone.class));
```
Comment (danikeichanka) [inline src/test/java/by/yurachel/web_app/servlet/AddPhoneServletTest.java:35]: ArgumentCaptor may be used here to verify which phone was added

File: src/test/java/by/yurachel/web_app/servlet/AdminServletTest.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    private AdminServlet adminServlet;
+
+    @BeforeEach
+    private void initialize() {
+        adminServlet = new AdminServlet();
+    }
+
+    @Test
+    void doGet() throws ServletException, IOException {
+        HttpServletRequest request = mock(HttpServletRequest.class);
```
Comment (danikeichanka) [inline src/test/java/by/yurachel/web_app/servlet/AdminServletTest.java:25]: Extend from HttpInit

File: src/test/java/by/yurachel/web_app/servlet/CatalogServletTest.java [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+        when(phoneRepository.getPhones()).thenReturn(phones);
+        catalogServlet.doGet(request, response);
+        verify(phoneRepository).getPhones(); // Проверяем,что у мока вызвался метод.
+        verify(request).setAttribute("phones", phoneRepository.getPhones());
+        verify(request).getRequestDispatcher("WEB-INF/catalog.jsp");
+        verify(requestDispatcher).forward(request, response);
+    }
+
+    @Test
+    void do_Get_No_Products_Returned() throws ServletException, IOException {
```
Comment (danikeichanka) [inline src/test/java/by/yurachel/web_app/servlet/CatalogServletTest.java:57]: Don't use underscore in the test name, just _testDoGetNoProductsReturned_

File: src/test/java/by/yurachel/web_app/servlet/LoginServletTest.java [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.*;
+
+
+class LoginServletTest extends HttpInit {
+    @InjectMocks
+    private LoginServlet loginServlet;
+    @Mock
+    private ServletConfig config;
```
Comment (danikeichanka) [inline src/test/java/by/yurachel/web_app/servlet/LoginServletTest.java:26]: Move to HttpInit as it may be requiredfor all servlets

**Pull request merged**